### PR TITLE
feat: permission driven CTAs for datasources

### LINKED ISSUE #16443: [Task]: Permission driven CTAs (hide/show)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

- create/delete/make public/export applications
-  create/delete pages 
- create/delete workspace 
- create/delete actions 
- create/delete/invite users/remove users  groups 
- create/delete roles

### LINKED ISSUE #16641: [Task]: Permission driven CTAs (hide/show) for create/delete for datasources

### PR DESCRIPTION
**Pull request by berzerkeer**
## Description
With the introduction of RBAC all resources have more fine grained access controls. This PR makes the resources create,delete and edit CTAs either hide/show based on the user permissions. 

Fixes #16443 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/ce/mocks/handlers.ts [hunk 1]
```diff
@@ -1,8 +1,47 @@
 import { rest } from "msw";
-import testMockApi from "./mockJsons/testMockApi.json";
```
Comment (ankitakinger) [inline app/client/src/ce/mocks/handlers.ts:2]: @berzerkeer revert changes of this file

File: app/client/src/pages/Applications/index.tsx [hunk 2]
```diff
@@ -393,6 +394,16 @@ function LeftPane() {
   const fetchedUserWorkspaces = useSelector(getUserApplicationsWorkspaces);
   const isFetchingApplications = useSelector(getIsFetchingApplications);
   const isMobile = useIsMobileDevice();
+
+  const userWorkspacePermissions = useSelector(
+    (state: AppState) => getCurrentAppWorkspace(state).userPermissions ?? [],
```
Comment (ankitakinger) [inline app/client/src/pages/Applications/index.tsx:399]: @berzerkeer we need to revert this to use permissions from the user key which we get in `/new` API

File: app/client/src/pages/Editor/Explorer/Widgets/WidgetContextMenu.tsx [hunk 3]
```diff
@@ -22,6 +22,8 @@ export function WidgetContextMenu(props: {
   widgetId: string;
   pageId: string;
   className?: string;
+  canCreatePages?: boolean;
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Widgets/WidgetContextMenu.tsx:25]: @berzerkeer is this being used here?

File: app/client/src/components/editorComponents/ActionRightPane/index.tsx [hunk 4]
```diff
@@ -275,7 +287,7 @@ function ActionSidebar({
           entityDependencies={entityDependencies}
         />
       )}
-      {hasResponse && Object.keys(widgets).length > 1 && (
+      {canEditPage && hasResponse && Object.keys(widgets).length > 1 && (
```
Comment (ankitakinger) [inline app/client/src/components/editorComponents/ActionRightPane/index.tsx:290]: @berzerkeer just check once if this needs to be just canEditPage check or it should also check if the action has edit action. I'm not sure, hence asking you to check.

File: app/client/src/pages/Editor/APIEditor/ApiAuthentication.tsx [hunk 5]
```diff
@@ -48,6 +53,7 @@ const OAuthText = styled.span<ErrorProps>`
 
 const DescriptionText = styled(Text)`
   margin: 12px auto;
+  ${(props) => props.disabled && "cursor: not-allowed;opacity:0.5"}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/APIEditor/ApiAuthentication.tsx:56]: @berzerkeer I think this can be removed as it is not being used.

File: app/client/src/components/editorComponents/ActionRightPane/index.tsx (hunk 4 above)
Comment (berzerkeer) [reply app/client/src/components/editorComponents/ActionRightPane/index.tsx:290]: I think it only needs canEditPage, since it just binds the data returned from the action to the widget on page. We are not editing the action AFAIK.

File: app/client/src/pages/Editor/APIEditor/ApiAuthentication.tsx (hunk 5 above)
Comment (berzerkeer) [reply app/client/src/pages/Editor/APIEditor/ApiAuthentication.tsx:56]: yes done.

File: app/client/src/ce/mocks/handlers.ts (hunk 1 above)
Comment (berzerkeer) [reply app/client/src/ce/mocks/handlers.ts:2]: should I remove the handler for the /workspaces api in this PR itself ?

File: app/client/src/pages/Applications/index.tsx (hunk 2 above)
Comment (berzerkeer) [reply app/client/src/pages/Applications/index.tsx:399]: done

File: app/client/src/ce/mocks/handlers.ts (hunk 1 above)
Comment (ankitakinger) [reply app/client/src/ce/mocks/handlers.ts:2]: Let it remain in this manner
image

Comment (berzerkeer) [reply app/client/src/ce/mocks/handlers.ts:2]: done

File: app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx [hunk 6]
```diff
@@ -101,75 +103,81 @@ export function ActionEntityContextMenu(props: EntityContextMenuProps) {
...
       defaultText=""
       modifiers={ContextMenuPopoverModifiers}
       onSelect={noop}
       optionTree={[
-        {
-          value: "rename",
-          onSelect: editActionName,
-          label: createMessage(CONTEXT_EDIT_NAME),
-        },
+        ...(props.canManageAction ? manageOptions : []),
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx:173]: @berzerkeer this will change the sequence of the action items. Can we instead do this:
1. Remove entire optionTree array outside return function and assign it to a variable
2. Add check to each object before itself.

File: app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx [hunk 7]
```diff
@@ -116,35 +133,38 @@ export function PageContextMenu(props: {
...
   return (
     <TreeDropdown
       className={props.className}
       defaultText=""
       modifiers={ContextMenuPopoverModifiers}
       onSelect={noop}
-      optionTree={optionTree}
+      optionTree={[
+        ...(canManagePages ? managePageOptions : []),
+        ...(canDeletePages ? [deletePageOption] : []),
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx:166]: @berzerkeer even though this is fine with what you have done in this component, can we aim at following a similar way of checking permissions for menu options all over?

Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx:166]: you mean just having it outside as one array and passing it ?

File: app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx (hunk 6 above)
Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx:173]: Ohh got it

File: app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx (hunk 7 above)
Comment (ankitakinger) [reply app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx:166]: yes

Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx]: done

File: app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx (hunk 6 above)
Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx]: done

File: app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx [hunk 8]
```diff
@@ -115,17 +132,7 @@ export function PageContextMenu(props: {
...
-    optionTree.push({
-      value: "setdefault",
-      onSelect: setPageAsDefaultCallback,
-      label: createMessage(CONTEXT_SET_AS_HOME_PAGE),
-    });
-  }
-
-  if (!props.isDefaultPage) {
-    optionTree.push({
+    canDeletePages && {
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Pages/PageContextMenu.tsx:135]: @berzerkeer we need this check here as well: `!props.isDefaultPage`. This existed before.

File: app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx [hunk 9]
```diff
@@ -124,12 +131,16 @@ class DatasourceDBEditor extends JSONtoForm<Props> {
...
+              <FormTitle
+                disabled={canManageDatasource}
+                focusOnMount={this.props.isNewDatasource}
+              />
             </FormTitleContainer>
             {viewMode && (
               <EditDatasourceButton
                 category={Category.tertiary}
                 className="t--edit-datasource"
+                // disabled={!canManageDatasource}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx:143]: @berzerkeer can we remove this comment?

File: app/client/src/pages/Editor/Explorer/Entity/index.tsx [hunk 10]
```diff
@@ -218,6 +220,7 @@ export type EntityProps = {
 
 export const Entity = forwardRef(
   (props: EntityProps, ref: React.Ref<HTMLDivElement>) => {
+    const { canEditEntityName, showAddButton = true } = props;
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Entity/index.tsx:223]: @berzerkeer can we assign `canEditEntityName` a default value of true?

File: app/client/src/pages/Editor/Explorer/EntityExplorer.tsx [hunk 11]
```diff
@@ -96,6 +98,12 @@ function EntityExplorer({ isActive }: { isActive: boolean }) {
     }
   }, [isFirstTimeUserOnboardingEnabled, pageId]);
 
+  const currentWorkspaceId = useSelector(getCurrentWorkspaceId);
+
+  useEffect(() => {
+    dispatch(fetchWorkspace(currentWorkspaceId));
+  }, [currentWorkspaceId]);
+
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/EntityExplorer.tsx:106]: @berzerkeer why are these changes needed?

File: app/client/src/pages/Editor/Explorer/JSActions/JSActionEntity.tsx [hunk 12]
```diff
@@ -61,13 +62,15 @@ export const ExplorerJSCollectionEntity = memo(
...
         active={props.isActive}
+        canEditEntityName={props.canManageActions}
         className="t--jsaction"
         contextMenu={contextMenu}
         entityId={jsAction.id}
         icon={JsFileIconV2}
         key={jsAction.id}
         name={jsAction.name}
         searchKeyword={props.searchKeyword}
+        showAddButton={false}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/JSActions/JSActionEntity.tsx:73]: @berzerkeer why is this false by default? Shouldn't this be `props.canCreateActions`?

File: app/client/src/pages/Editor/Explorer/Actions/ActionEntity.tsx [hunk 13]
```diff
@@ -68,6 +86,7 @@ export const ExplorerActionEntity = memo((props: ExplorerActionEntityProps) => {
     <Entity
       action={switchToAction}
       active={props.isActive}
+      canEditEntityName={canManageAction}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Actions/ActionEntity.tsx:89]: @berzerkeer don't we need `showAddButton` prop in this one?

File: app/client/src/pages/Editor/Explorer/Files/index.tsx [hunk 14]
```diff
@@ -67,6 +90,7 @@ function Files() {
         } else if (type === "JS") {
           return (
             <ExplorerJSCollectionEntity
+              canManageActions={canManageActions}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Files/index.tsx:93]: @berzerkeer why not follow the same prop passing logic in the else block instead of fetching permissions again in that component?

File: app/client/src/pages/Editor/Explorer/Pages/index.tsx [hunk 15]
```diff
@@ -202,6 +226,7 @@ function Pages() {
             preRightIcon={isCurrentPage ? currentPageIcon : ""}
             rightIcon={rightIcon}
             searchKeyword={""}
+            showAddButton={canCreatePages}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/Explorer/Pages/index.tsx:229]: @berzerkeer this is not needed as we do not show add button next to each page already added.

File: app/client/src/pages/Editor/Explorer/EntityExplorer.tsx (hunk 11 above)
Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/EntityExplorer.tsx:106]: Because we need workspace permissions in store for datasources, so once we open the app, we are fetching workspace so that it updates the workspace permissions in store.

File: app/client/src/pages/Editor/Explorer/JSActions/JSActionEntity.tsx (hunk 12 above)
Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/JSActions/JSActionEntity.tsx:73]: We don't need this at all, i forgot to remove it. This represents each individual JS object.

File: app/client/src/pages/Editor/Explorer/Pages/index.tsx (hunk 15 above)
Comment (berzerkeer) [reply app/client/src/pages/Editor/Explorer/Pages/index.tsx:229]: you are correct, removed.

File: app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx [hunk 16]
```diff
@@ -176,7 +196,7 @@ function ContextMenu(props: Props) {
           <main>
             <h4>General</h4>
             {!page.isDefault && (
-              <MenuItem>
+              <MenuItem aria-disabled={!canManagePages}>
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/PagesEditor/ContextMenu.tsx:199]: @berzerkeer None of these seems to be working as even though the icons are disabled, the click works. Can we re-check this file please?

File: app/client/src/pages/Editor/PagesEditor/PageListItem.tsx [hunk 17]
```diff
@@ -200,6 +217,7 @@ function PageListItem(props: PageListItemProps) {
                 <Action type="button">
                   <CopyIcon
                     color={Colors.GREY_9}
+                    disabled={!canManagePages}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/PagesEditor/PageListItem.tsx:220]: @berzerkeer we need to re-check this one too. These also work even if disabled.

File: app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx [hunk 18]
```diff
@@ -105,13 +112,17 @@ class DatasourceSaaSEditor extends JSONtoForm<Props> {
...
+                disabled={canManageDatasource}
+                focusOnMount={this.props.isNewDatasource}
+              />
             </FormTitleContainer>
 
             {viewMode && (
               <EditDatasourceButton
                 category={Category.tertiary}
                 className="t--edit-datasource"
+                // disabled={!canManageDatasource}
```
Comment (ankitakinger) [inline app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx:125]: @berzerkeer can we remove this comment?
